File: app/src/main/java/com/chuongvd/app/signal/data/source/remote/service/ErrorHandlerFactory.java
package com.chuongvd.app.signal.data.source.remote.service;

import android.util.Log;
import com.google.gson.Gson;
import com.google.gson.JsonSyntaxException;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.annotations.NonNull;
import io.reactivex.annotations.Nullable;
import io.reactivex.functions.Function;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.HttpException;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

public class ErrorHandlerFactory extends CallAdapter.Factory {
    private final RxJava2CallAdapterFactory mInstance;

    private ErrorHandlerFactory() {
        mInstance = RxJava2CallAdapterFactory.create();
    }

    @Nullable
    @Override
    public CallAdapter<?, ?> get(Type type, Annotation[] annotations, Retrofit retrofit) {
        return new RxCallAdapterWrapper(retrofit, mInstance.get(type, annotations, retrofit));
    }

    public static CallAdapter.Factory create() {
        return new ErrorHandlerFactory();
    }

    private class RxCallAdapterWrapper<R> implements CallAdapter<R, Observable<?>> {

        final Retrofit mRetrofit;
        final CallAdapter<R, Observable<?>> mWrapped;

        public RxCallAdapterWrapper(Retrofit retrofit, CallAdapter<R, Observable<?>> callAdapter) {

            this.mRetrofit = retrofit;
            this.mWrapped = callAdapter;
        }

        @Override
        public Type responseType() {
            return mWrapped.responseType();
        }

        @SuppressWarnings("unchecked")
        @Override
        public Observable<?> adapt(Call<R> call) {
            return ((Observable) mWrapped.adapt(call)).onErrorResumeNext(
                    new Function<Throwable, ObservableSource>() {
                        @Override
                        public ObservableSource apply(@NonNull Throwable throwable)
                                throws Exception {
                            if (throwable == null) {
                                return Observable.error(throwable);
                            }
                            return Observable.error(convertToBaseException(throwable));
                        }
                    });
        }

        private BaseException convertToBaseException(Throwable throwable) {
            if (throwable instanceof BaseException) {
                return (BaseException) throwable;
            } else if (throwable instanceof IOException) {
                return BaseException.toNetworkError(throwable);
            } else if (throwable instanceof HttpException) {
                HttpException httpException = (HttpException) throwable;
                Response response = httpException.response();
                if (response.errorBody() == null) {
                    return BaseException.toHttpError(response);
                }
                try {
                    String errorResponse = response.errorBody().string();
                    BaseErrorResponse baseErrorResponse = deserializeErrorBody(errorResponse);
                    if (baseErrorResponse != null && baseErrorResponse.isError()) {
                        //Get error data from Server
                        return BaseException.toServerError(baseErrorResponse);
                    } else {
                        //Get error data cause http connection
                        return BaseException.toHttpError(response, baseErrorResponse);
                    }
                } catch (IOException e) {
                    Log.e(this.getClass().getSimpleName(), e.getMessage());
                }
            }
            return BaseException.toUnexpectedError(throwable);
        }

        private BaseErrorResponse deserializeErrorBody(String errorString) {
            try {
                return new Gson().fromJson(errorString, BaseErrorResponse.class);
            } catch (JsonSyntaxException e) {
                return null;
            }
        }
    }
}

File: app/src/main/java/com/chuongvd/app/signal/data/source/remote/service/ErrorHandlerFactory.kt
//package com.chuongvd.app.signal.data.source.remote.service
//
//import android.util.Log
//import com.google.gson.Gson
//import com.google.gson.JsonSyntaxException
//import io.reactivex.Observable
//import io.reactivex.ObservableSource
//import io.reactivex.annotations.Nullable
//import io.reactivex.functions.Function
//import retrofit2.Call
//import retrofit2.CallAdapter
//import retrofit2.HttpException
//import retrofit2.Retrofit
//import java.io.IOException
//import java.lang.reflect.Type
//
//class ErrorHandlerFactory private constructor() : CallAdapter.Factory() {
//  private val mInstance: ErrorHandlerFactory
//
//  init {
//    mInstance = ErrorHandlerFactory()
//  }
//
//  @Nullable
//  override operator fun get(type: Type, annotations: Array<Annotation>,
//      retrofit: Retrofit): CallAdapter<*, Observable<*>> {
//    return RxCallAdapterWrapper(retrofit, mInstance.get(type, annotations, retrofit))
//  }
//
//  private inner class RxCallAdapterWrapper<R>(internal val mRetrofit: Retrofit,
//      internal val mWrapped: CallAdapter<R, Observable<*>>) : CallAdapter<R, Observable<*>> {
//
//    override fun adapt(call: Call<R>?): Observable<*> {
//    }
//
//    override fun responseType(): Type {
//      return mWrapped.responseType()
//    }
//
////    override fun adapt(call: Call<R>): Observable<*> {
////      return (mWrapped.adapt(call) as Observable<*>).onErrorResumeNext(
////          object : Function<Throwable, ObservableSource<*>> {
////            override fun apply(throwable: Throwable): ObservableSource<*> {
////              return Observable.error<Any>(convertToBaseException(throwable))
////            }
////          })
////    }
//
//    private fun convertToBaseException(throwable: Throwable): BaseException {
//      if (throwable is BaseException) {
//        return throwable
//      } else if (throwable is IOException) {
//        return BaseException.toNetworkError(throwable)
//      } else if (throwable is HttpException) {
//        val httpException = throwable as HttpException?
//        val response = httpException!!.response()
//        if (response.errorBody() == null) {
//          return BaseException.toHttpError(response)
//        }
//        try {
//          val errorResponse = response.errorBody()!!.string()
//          val baseErrorResponse = deserializeErrorBody(errorResponse)
//          return if (baseErrorResponse != null && baseErrorResponse.isError) {
//            //Get error data from Server
//            BaseException.toServerError(baseErrorResponse)
//          } else {
//            //Get error data cause http connection
//            baseErrorResponse?.let { BaseException.toHttpError(response, it) }!!
//          }
//        } catch (e: IOException) {
//          Log.e(this.javaClass.simpleName, e.message)
//        }
//
//      }
//      return BaseException.toUnexpectedError(throwable!!)
//    }
//
//    private fun deserializeErrorBody(errorString: String): BaseErrorResponse? {
//      try {
//        return Gson().fromJson(errorString, BaseErrorResponse::class.java)
//      } catch (e: JsonSyntaxException) {
//        return null
//      }
//
//    }
//  }
//
//  companion object {
//
//    fun create(): CallAdapter.Factory {
//      return ErrorHandlerFactory()
//    }
//  }
//}